### PR TITLE
add support for constant

### LINKED ISSUE #58739: Python Array API: support constants
The [array API specification stipulates](https://data-apis.org/array-api/latest/API_specification/constants.html) four numeric constants:

- `e`
- `inf`
- `nan`
- `pi`

PyTorch currently supports none.

cc @mruberry @rgommers @pmeier

### PR DESCRIPTION
**Pull request by praneethratna**
Fixes #58739 Add support for constants according to python array API stipulation.

### CODE REVIEW

**Comment by praneethratna** [comment]
@pmeier sorry as I have made a mess with commits , so have to delete the previous branch, I`ve updated here again

**Review by pmeier — CHANGES_REQUESTED** [review]
> sorry as I have made a mess with commits , so have to delete the previous branch, I`ve updated here again

Happens to the best of us :wink: You accidentally included updates to the `ideep` submodule in the branch. Here is how you get them out again:

```shell
$ cd $PYTORCH/third_party/ideep 
$ git checkout f9468ff1a3d601b509ebe2c17d2ed0a58dffacee
$ git commit -am "revert submodule update"
$ git push
```

If you run `git submodule update` after doing a `git pull` you won't have this problem in the future.

**Comment by pmeier** [comment]
For reference, the old PR is #59910.

**Comment by praneethratna** [comment]
@pmeier I have tried your steps but i am getting a permission denied while trying to push could you help me figure it out?

**Comment by pmeier** [comment]
You need to perform these steps while on your branch. "Permission denied" most likely means that your trying to push to branch that isn't yours. Probably master.

**Comment by praneethratna** [comment]
I have tried these steps on my branch itself but still it shows "Permission Denied"

**Comment by pmeier** [comment]
Ahh sorry, do a `cd $PYTORCH` before the commit.

**Comment by pmeier** [comment]
The commit is still there. Is it ok for you if I fix the branch so we can land this PR?

**Comment by praneethratna** [comment]
yes absolutely not an issue 😃

**Review by pmeier — APPROVED** [review]
Should be good to go now. Thanks @praneethratna!

File: torch/__init__.py [hunk 1]
```diff
@@ -511,6 +511,14 @@ def is_warn_always_enabled():
     """
     return _C._get_warnAlways()
 
+################################################################################
+# Define numeric constants
+################################################################################
+
+# for Python Array API consistency (https://data-apis.org/array-api/latest/API_specification/constants.html) and NumPy consistency (https://numpy.org/devdocs/reference/constants.html)
```
Comment (pmeier) [inline torch/__init__.py:518]:
```suggestion
# For Python Array API (https://data-apis.org/array-api/latest/API_specification/constants.html) and 
# NumPy consistency (https://numpy.org/devdocs/reference/constants.html)
```

File: torch/__init__.py [hunk 2]
```diff
@@ -511,6 +511,15 @@ def is_warn_always_enabled():
     """
     return _C._get_warnAlways()
 
+################################################################################
+# Define numeric constants
+################################################################################
+
+# For Python Array API (https://data-apis.org/array-api/latest/API_specification/constants.html) and 
+# NumPy consistency (https://numpy.org/devdocs/reference/constants.html)
```
Comment (pmeier) [inline torch/__init__.py:519]:
```suggestion
# For Python Array API (https://data-apis.org/array-api/latest/API_specification/constants.html) and
# NumPy consistency (https://numpy.org/devdocs/reference/constants.html)
```

**Review by mruberry — APPROVED** [review]
Cool, thank you @praneethratna!

**Review by mikaylagawarecki — COMMENTED** [review]
EDIT: Sorry I unintentionally commented here

**Pull request closed without merging**